substrate/client/service/src/client/client.rs

```diff
@@ -128,6 +128,10 @@ where
 	fn clear_block_gap(&self) -> sp_blockchain::Result<()> {
 		self.backend.blockchain().clear_block_gap()
 	}
+
+	fn finalize_block(&self, hash: Block::Hash) -> sp_blockchain::Result<()> {
```

Comment:
There is already `apply_finality` in Substrate, why do we have to add another API on top of this instead of fixing the API we already have?

Reply:
Indeed, `apply_finality` relies on the `tree_route` which demands existing headers from the genesis in our case. Changing the method seems much more invasive than preparing conditions for its correct work - I'm open to suggestions here though.